### PR TITLE
[IMP] hw{_drivers,_posbox_homepage}: use python lib to connect Wi-Fi

### PR DESCRIPTION
**Pull request by loouis-t**
The IoT Box used multiple hardly maintainable Bash scripts to connect to Wi-Fi.

This commit brings a `wifi.py` file including all the methods needed to abstract the use of `nmcli` (`NetworkManager` CLI) to easily manage Wi-Fi configuration directly from python.

Task: 4166329

### CODE REVIEW

**Comment by loouis-t** [comment]
@odoo/rd-security could you have a look at this please? This PR is a PR we did in `17.0` that we retargeted in master  (https://github.com/odoo/odoo/pull/179341).
I applied the changes @xmo-odoo requested in `17.0`.

**Review by xmo-odoo — CHANGES_REQUESTED** [review]
@robodoo override=ci/security

File: addons/hw_drivers/tools/wifi.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+
+def is_access_point():
+    """Check if the device is currently in access point mode.
+
+    :return: True if the device is in access point mode
+    :rtype: bool
+    """
+    # We only get the first line as `loopback` can still be active when `wlan0` is up
+    device = _nmcli(['-g', 'device', 'connection', 'show', '--active'], sudo=True).splitlines()[0]
+    return device == 'loopback'
```
Comment (MaxTheWhale) [inline addons/hw_drivers/tools/wifi.py:278]: This is always returning `False` for me (testing on RPi 5). When I run the command manually it returns `lo` not `loopback`.

File: addons/hw_drivers/tools/wifi.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+import logging
+import secrets
+import subprocess
+import time
+from pathlib import Path
+from functools import cache
+
+from .helpers import get_ip, get_mac_address, writable
+
+__all__ = [
```
Comment (MaxTheWhale) [inline addons/hw_drivers/tools/wifi.py:14]: This is the first time I've seen `__all__`, from what I've read it is only needed when you have `import *` - what was the reason for adding it?

Comment (loouis-t) [reply addons/hw_drivers/tools/wifi.py:14]: I can remove it, I added it in case someone, one day, decide to `import *`, it's kind of a "security" to avoid people to use methods they shouldn't. If you think it's useless, let's remove it

File: addons/hw_drivers/tools/wifi.py (hunk 1 above)
Comment (loouis-t) [reply addons/hw_drivers/tools/wifi.py:278]: ah yes you're right, sorry

File: addons/hw_drivers/tools/wifi.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    :return: list of found SSIDs with a flag indicating whether it's the connected network
+    :rtype: list[tuple[bool, str]]
+    """
+    ssids = _nmcli(['-f', 'ACTIVE,SSID', 'dev', 'wifi'], sudo=True)
+    ssids_dict = {
+        ssid.split(':')[-1]: ssid.startswith('yes:')
+        for ssid in sorted(ssids.splitlines())
+        if ssid
+    } if ssids else {}
+
```
Comment (Ysoroko) [inline addons/hw_drivers/tools/wifi.py:50]: add a logger debug

**Review by Ysoroko — APPROVED** [review]
Everything works:
1) Access point
2) Wifi
3) Wifi reconnection on reboot (persistent)
4) Wifi disconnection

I would add some additional loggers but otherwise it looks good to me

File: addons/hw_posbox_homepage/static/src/app/components/dialog/WifiDialog.js [hunk 4]
```diff
@@ -17,11 +17,11 @@ export class WifiDialog extends Component {
             loading: false,
             waitRestart: false,
             availableWifi: [],
+            currentWifi: '',
```
Comment (Ysoroko) [inline addons/hw_posbox_homepage/static/src/app/components/dialog/WifiDialog.js:20]: Looks like this variable only has one value and is unused

File: addons/iot_box_image/overwrite_before_init/etc/init_image.sh [hunk 5]
```diff
@@ -221,6 +221,7 @@ groupadd usbusers
 usermod -a -G usbusers odoo
 usermod -a -G lp odoo
 usermod -a -G input lightdm
+usermod -a -G pi odoo
```
Comment (loouis-t) [inline addons/iot_box_image/overwrite_before_init/etc/init_image.sh:224]: allow `odoo` access `/etc/hostapd/hostapd.conf` as it belongs to `pi` user and group.
Will also be a safety for `odoo` user to be able to access all `pi` files

**Review by Ysoroko — APPROVED** [review]
@robodoo r+

**Comment by qle-odoo** [comment]
This PR work with windows compatibility ?

**Comment by loouis-t** [comment]
@robodoo r-

File: addons/hw_drivers/main.py [hunk 6]
```diff
@@ -80,8 +80,13 @@ def send_alldevices(self, iot_client=None):
...
     def run(self):
-        """Thread that will load interfaces and drivers and contact the odoo server with the updates"""
-        self.server_url = helpers.get_odoo_server_url()
+        """Thread that will load interfaces and drivers and contact the odoo server
+        with the updates. It will also reconnect to the Wi-Fi if the connection is lost.
+        """
+        if platform.system() == 'Linux':
+            wifi.reconnect(helpers.get_conf('wifi_ssid'), helpers.get_conf('wifi_password'))
+        # Only contact db if connected to the internet
+        self.server_url = helpers.get_odoo_server_url() if wifi.get_current() else None
```
Comment (Ysoroko) [inline addons/hw_drivers/main.py:89]: wifi.get_current() isn't working on Windows

File: addons/hw_drivers/tools/helpers.py [hunk 7]
```diff
@@ -286,25 +284,15 @@ def get_mac_address():
...
 def get_odoo_server_url():
     """Get the URL of the linked Odoo database.
-    If the IoT Box is in access point mode, it will return ``None`` to avoid
-    connecting to the server.
 
     :return: The URL of the linked Odoo database.
     :rtype: str or None
     """
-    return None if access_point() else get_conf('remote_server')
+    return get_conf('remote_server')
```
Comment (Ysoroko) [inline addons/hw_drivers/tools/helpers.py:295]: in case of an access point configuration this change will potentially provoke a lot of request errors

File: addons/point_of_sale/tools/posbox/.gitignore [hunk 8]
```diff
@@ -1,4 +0,0 @@
-raspios.img
-iotbox.img
-root_mount/
-overwrite_before_init/usr/bin/ngrok
```
Comment (Ysoroko) [inline addons/point_of_sale/tools/posbox/.gitignore:4]: put it in a separate commit

**Comment by Ysoroko** [comment]
@robodoo r+
